optimum/habana/trl/trainer/ddpo_trainer.py

```diff
@@ -0,0 +1,506 @@
+                    "log_probs": log_probs,
+                    "negative_prompt_embeds": sample_neg_prompt_embeds,
+                }
+            )
+            prompt_image_pairs.append([images, prompts, prompt_metadata])
+
+        return samples, prompt_image_pairs
+
+    def _train_batched_samples(self, inner_epoch, epoch, global_step, batched_samples):
+        """
+        Train on a batch of samples. Main training segment
+
+        Adapted from https://github.com/huggingface/trl/blob/v0.7.6/trl/trainer/ddpo_trainer.py#L483
+        - Reduce recompilations by avoiding constant variables in loops
+        - Add `mark_step()` to support lazy mode
+
+        Args:
+            inner_epoch (int): The current inner epoch
+            epoch (int): The current epoch
+            global_step (int): The current global step
+            batched_samples (List[Dict[str, torch.Tensor]]): The batched samples to train on
+
+        Side Effects:
+            - Model weights are updated
+            - Logs the statistics to the accelerator trackers.
+
+        Returns:
+            global_step (int): The updated global step
+        """
+        info = defaultdict(list)
+
+        for i, sample in enumerate(batched_samples):
+            if self.config.train_cfg:
+                # concat negative prompts to sample prompts to avoid two forward passes
+                embeds = torch.cat([sample["negative_prompt_embeds"], sample["prompt_embeds"]])
+            else:
+                embeds = sample["prompt_embeds"]
+
+            latents = sample["latents"]
+            timesteps = sample["timesteps"]
+            next_latents = sample["next_latents"]
+            log_probs = sample["log_probs"]
+
+            for j in range(self.num_train_timesteps): #, desc=f"Epoch{i}"):
+                with self.accelerator.accumulate(self.sd_pipeline.unet):
+                    # Reduce recompilations by avoiding constant variables in loops
+                    latent = latents[:, 0]
+                    timestep = timesteps[:, 0]
+                    next_latent = next_latents[:, 0]
+                    log_prob = log_probs[:, 0]
+                    latents = torch.roll(latents, shifts=-1, dims=1)
+                    timesteps = torch.roll(timesteps, shifts=-1, dims=1)
+                    next_latents = torch.roll(next_latents, shifts=-1, dims=1)
+                    log_probs = torch.roll(log_probs, shifts=-1, dims=1)
+
+                    loss, approx_kl, clipfrac = self.calculate_loss(
+                        latent,
+                        timestep,
+                        next_latent,
+                        log_prob,
+                        sample["advantages"],
+                        embeds,
+                    )
+
+                    info["approx_kl"].append(approx_kl)
+                    info["clipfrac"].append(clipfrac)
+                    info["loss"].append(loss)
+
+                    self.accelerator.backward(loss)
```

Comment:
need a mark_step() after backward also.

Reply:
Thanks @yeonsily  I'll add that

Reply:
@yeonsily I added mark_step() after backward